<html>

  <head>

    <title>NCBI VDB-2: Repository</title>
    <link href="vdb-2.css" rel="stylesheet" type="text/css"/>

  </head>

  <body>

    <!-- Document Header -->

    <h1>VDB-2 - Repository</h1>

    <p class="toc">
    <span class="hdr">Type:</span><br/>
      <span class="entry">component</span>
    </p>

    <p class="toc">
      <span class="hdr">Header:</span><br/>
      <span class="entry"><a href="../interfaces/kfg/repository.h">&lt;kfg/repository.h&gt;</a></span>
    </p>

    <p class="toc">
    <span class="hdr">Revision History:</span><br/>
      <table class="entry">
        <tr><td>2013-Nov-12</td><td>&bull;</td><td>rodarmer</td></tr>
      </table>
    </p>

    <p class="toc">
      <span class="hdr">Contents:</span><br/>
      <ul class="toc-outer">
        <li><a href="#abstract">Abstract</a></li>
        <li><a href="#purpose">Purpose</a></li>
        <li><a href="#history">History</a></li>
        <li><a href="#description">Description</a></li>
        <li><a href="#requirements">Requirements</a></li>
        <li><a href="#implementation">Implementation</a></li>
      </ul>
    </p>

    <hr/>

    <!-- Abstract -->

    <a name="abstract">

    <h2>Abstract</h2>

    <div class="section">

	  <p>
		The VDB <strong>Repository</strong> object embodies physical storage used
		within a multi-tiered distributed archive.
	  </p>

	</div>

	</a>

    <!-- Purpose -->

    <a name="purpose">

    <h2>Purpose</h2>

    <div class="section">

	  <p>
		A <em>Repository</em> supports declaration of a cache as a unit within a
		multi-tier archive. This in turn supports local caching of downloaded data
		for the general case that downloads are considered more expensive than
		local storage. It also allows establishment of one or more organization-wide
		<em>Repositories</em> to cache public data on a large scale for several users.
	  </p>

	  <p>
		The typical configuration will have a single root node at NCBI, accessed
		remotely via <strong>HTTP</strong>, and a single leaf node in the user's
		home directory representing full and partial downloads in the form of a
		local cache. Within NCBI, we have an unique configuration where there is
		no <em>remote</em> root node, but instead the master archive itself serves
		as a <strong>site</strong> <em>Repository</em>, and users do not have their
		own caches.
	  </p>

	  <p>
		We are going to introduce a higher notion of <strong>FileCache</strong> such
		that a cache can be comprised of 1+ <em>Repositories</em>. A <em>cache</em>
		will either be <strong>public</strong> or associated with a
		<strong>dbGaP consent group</strong>. So the <em>Repository</em> is used as
		a sub-component of the <em>FileCache</em>.
	  </p>

	  <p>
		We are also going to introduce the notion of <strong>Workspace</strong> that
		designates where work should be done. It will have a <em>Repository</em> or a
		<em>FileCache</em> as a sub-component describing where to store results.
	  </p>

	  <p>
		Finally, we are going to introduce the <strong>Project</strong> or something
		on those lines that will identify a set of files upon which to work, as imported
		from one or more cart files, etc. A <em>Project</em> establishes a namespace,
		and designates a <em>Workspace</em> for operations. The sources are associated
		with a <em>FileCache</em> for inputs.
	  </p>

	</div>

	</a>

    <!-- History -->

    <a name="history">

    <h2>History</h2>

    <div class="section">

	  <p>
		The <em>Repository</em> grew out of an NCBI-internal description of file servers
		used in the <strong>Trace</strong> project. The file server was described with a
		<strong>root path</strong>, and each file sever contained a set of
		<strong>volumes</strong> for storing <em>Trace</em> database objects.
	  </p>

	  <p>
		<em>Trace</em> objects were treated as movable across the volumes, a feature
		that prevented mismatches between name-resolution databases and the file system.
		Under this scheme, dynamic name-resolution was performed on each attempt to
		open an object, such that each volume upon each server was searched until the
		object was found. The format upon each volume was based upon object accession,
		where each accession would be transformed into a canonical partial path,
		reducing the <em>search</em> to a single test per volume.
	  </p>

	  <p>
		This approach was applied to the <strong>SRA</strong> archive, nearly
		verbatim at the beginning. Later, partners within INSDC diverged in their
		canonical path creation to allow a repository to be browsed, changing the
		computer-friendly 1024-based binning to human-friendly 1000-based binning.
		To handle this divergence, each volume was given a name-expansion algorithm,
		initially based upon the INSDC parter abbreviations <strong>ncbi</strong>,
		<strong>ebi</strong>, and <strong>ddbj</strong>.
	  </p>

	  <p>
		At present <em>(11/12/2013)</em> we have 16 volumes dedicated to NCBI,
		8 to EBI, and 1 to DDBJ in our public repository. We have extended the idea
		of <em>Repository</em> to encompass tiers <strong>remote</strong>,
		<strong>site</strong>, and <strong>user</strong>. Each tier is divided into
		sub-categories as appropriate for the tier:
	  </p>

	  <ul>
		<li><em>remote</em> - <strong>main</strong>, <strong>aux</strong> and <strong>protected</strong></li>
		<li><em>site</em> - <strong>main</strong> and <strong>aux</strong></li>
		<li><em>user</em> - <strong>public</strong> and <strong>protected</strong></li>
	  </ul>

	</div>

	</a>

    <!-- Description -->

    <a name="description">

    <h2>Description</h2>

    <div class="section">

	  <p>
		A <strong>Repository</strong> supports declaration of a cache upon a file server.
		It is categorized by a three-level key as described above:
		<strong>&lt;category&gt;&lt;sub-category&gt;&lt;name&gt;</strong>.
	  </p>

	  <p>
		Each named <em>Repository</em> is divided into several logical partitions by data
		classification/application. These relate to cache design and are known to the
		name resolver:
	  </p>

	  <ul>
		<li><strong>any</strong> - a categorization that serves as a wildcard</li>
		<li><strong>file</strong> - phenotype and genotype files</li>
		<li><strong>refseq</strong> - extern reference sequences</li>
		<li><strong>sra</strong> - SRA sequences and alignments</li>
		<li><strong>wgs</strong> - WGS contigs and scaffolds</li>
		<li><strong>nannot</strong> - named annotations</li>
		<li><strong>nakmer</strong> - named kmers</li>
	  </ul>

	  <p>
		Each application partition is allowed one or more volume name expansion
		algorithms, gathered under the <strong>&lt;volumes&gt;</strong> key node:
	  </p>

	  <ul>
		<li><strong>cgi</strong> - send request to NCBI for resolution</li>
		<li><strong>flat</strong> - use name or accession <q>as-is</q></li>
		<li><strong>sraFlat</strong> - append <q>.sra</q> to accession</li>
		<li><strong>sra1024</strong> - expand accession into 3 part path using 1024 entry banks</li>
		<li><strong>sra1000</strong> - expand accession into 3 part path using 1000 entry banks</li>
		<li><strong>fuse1000</strong> - expand accession into 4 part path using 1000 entry banks</li>
		<li><strong>refseq</strong> - use name <q>as-is</q> unless WGS</li>
		<li><strong>wgsFlat</strong> - use name <q>as-is</q>, potentially add extension in future</li>
		<li><strong>wgs</strong> - expand name into 4 part path</li>
		<li><strong>fuseWGS</strong> - expand name into 3 part path</li>
		<li><em>more to be added...</em></li>
	  </ul>

	</div>

	</a>

  </body>

</html>
